File: eco-core/core-plugin/src/main/java/com/willfp/reforges/reforges/util/ReforgeHandler.java
package com.willfp.reforges.reforges.util;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.PluginDependent;
import com.willfp.eco.core.gui.menu.Menu;
import com.willfp.eco.core.gui.slot.Slot;
import com.willfp.eco.core.integrations.economy.EconomyManager;
import com.willfp.reforges.reforges.PriceMultipliers;
import com.willfp.reforges.reforges.Reforge;
import com.willfp.reforges.reforges.meta.ReforgeTarget;
import org.bukkit.Sound;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class ReforgeHandler extends PluginDependent<EcoPlugin> {
    /**
     * Pass an {@link EcoPlugin} in order to interface with it.
     *
     * @param plugin The plugin to manage.
     */
    public ReforgeHandler(@NotNull EcoPlugin plugin) {
        super(plugin);
    }

    public void handleReforgeClick(@NotNull final InventoryClickEvent event,
                                   @NotNull final Slot slot,
                                   @NotNull final Menu menu) {
        Player player = (Player) event.getWhoClicked();
        ItemStack toReforge = menu.getCaptiveItems(player).isEmpty() ? null : menu.getCaptiveItems(player).get(0);
        if (toReforge == null) {
            return;
        }

        Reforge existingReforge = ReforgeUtils.getReforge(toReforge);

        List<ReforgeTarget> target = ReforgeTarget.getForItem(toReforge);

        Reforge reforge = null;
        boolean usedStone = false;

        if (menu.getCaptiveItems(player).size() == 2) {
            Reforge stone = ReforgeUtils.getReforgeStone(menu.getCaptiveItems(player).get(1));
            if (stone != null) {
                if (stone.getTargets().stream().anyMatch(reforgeTarget -> reforgeTarget.matches(toReforge))) {
                    reforge = stone;
                    usedStone = true;
                }
            }
        }

        if (reforge == null) {
            List<Reforge> existing = new ArrayList<>();
            if (existingReforge != null) {
                existing.add(existingReforge);
            }
            reforge = ReforgeUtils.getRandomReforge(target, existing);
        }

        if (reforge == null) {
            return;
        }

        double cost = 0;
        if (EconomyManager.hasRegistrations()) {
            cost = this.getPlugin().getConfigYml().getDouble("reforge.cost");
            int reforges = ReforgeUtils.getReforges(toReforge);
            cost *= Math.pow(this.getPlugin().getConfigYml().getDouble("reforge.cost-exponent"), reforges);
            if (reforge.getRequiresStone() && reforge.getStonePrice() != -1) {
                cost = reforge.getStonePrice();
            }
            cost *= PriceMultipliers.getForPlayer(player).getMultiplier();

            if (!EconomyManager.hasAmount(player, cost)) {
                player.sendMessage(this.getPlugin().getLangYml().getMessage("insufficient-money"));

                if (this.getPlugin().getConfigYml().getBool("gui.insufficient-money-sound.enabled")) {
                    player.playSound(
                            player.getLocation(),
                            Sound.valueOf(this.getPlugin().getConfigYml().getString("gui.insufficient-money-sound.id").toUpperCase()),
                            1f,
                            (float) this.getPlugin().getConfigYml().getDouble("gui.insufficient-money-sound.pitch")
                    );
                }

                return;
            }
        }

        int xpCost = this.getPlugin().getConfigYml().getInt("reforge.xp-cost");
        int reforges = ReforgeUtils.getReforges(toReforge);
        xpCost *= Math.pow(this.getPlugin().getConfigYml().getDouble("reforge.cost-exponent"), reforges);
        xpCost *= PriceMultipliers.getForPlayer(player).getMultiplier();
        if (player.getLevel() < xpCost) {
            player.sendMessage(this.getPlugin().getLangYml().getMessage("insufficient-xp"));
            if (this.getPlugin().getConfigYml().getBool("gui.insufficient-money-sound.enabled")) {
                player.playSound(
                        player.getLocation(),
                        Sound.valueOf(this.getPlugin().getConfigYml().getString("gui.insufficient-money-sound.id").toUpperCase()),
                        1f,
                        (float) this.getPlugin().getConfigYml().getDouble("gui.insufficient-money-sound.pitch")
                );
            }

            return;
        }

        if (EconomyManager.hasRegistrations()) {
            EconomyManager.removeMoney(player, cost);
        }

        player.setLevel(player.getLevel() - xpCost);

        player.sendMessage(this.getPlugin().getLangYml().getMessage("applied-reforge").replace("%reforge%", reforge.getName()));

        ReforgeUtils.incrementReforges(toReforge);

        ReforgeUtils.setReforge(toReforge, reforge);

        if (usedStone) {
            ItemStack stone = menu.getCaptiveItems(player).get(1);
            stone.setItemMeta(null);
            stone.setAmount(0);

            if (this.getPlugin().getConfigYml().getBool("gui.stone-sound.enabled")) {
                player.playSound(
                        player.getLocation(),
                        Sound.valueOf(this.getPlugin().getConfigYml().getString("gui.stone-sound.id").toUpperCase()),
                        1f,
                        (float) this.getPlugin().getConfigYml().getDouble("gui.stone-sound.pitch")
                );
            }

        }

        if (this.getPlugin().getConfigYml().getBool("gui.sound.enabled")) {
            player.playSound(
                    player.getLocation(),
                    Sound.valueOf(this.getPlugin().getConfigYml().getString("gui.sound.id").toUpperCase()),
                    1f,
                    (float) this.getPlugin().getConfigYml().getDouble("gui.sound.pitch")
            );
        }

    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/reforges/util/ReforgeHandler.kt
package com.willfp.reforges.util

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.gui.menu.Menu
import com.willfp.eco.core.integrations.economy.EconomyManager
import com.willfp.reforges.reforges.PriceMultipliers.getForPlayer
import com.willfp.reforges.reforges.Reforge
import com.willfp.reforges.reforges.ReforgeTarget.Companion.getForItem
import com.willfp.reforges.util.ReforgeUtils.getRandomReforge
import com.willfp.reforges.util.ReforgeUtils.getReforge
import com.willfp.reforges.util.ReforgeUtils.getReforgeStone
import com.willfp.reforges.util.ReforgeUtils.getReforges
import com.willfp.reforges.util.ReforgeUtils.incrementReforges
import com.willfp.reforges.util.ReforgeUtils.setReforge
import org.bukkit.Sound
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent
import java.util.*
import kotlin.math.pow

class ReforgeHandler(private val plugin: EcoPlugin) {
    fun handleReforgeClick(
        event: InventoryClickEvent,
        menu: Menu
    ) {
        val player = event.whoClicked as Player
        val toReforge =
            (if (menu.getCaptiveItems(player).isEmpty()) null else menu.getCaptiveItems(player)[0]) ?: return
        val existingReforge = getReforge(toReforge)
        val target = getForItem(toReforge)
        var reforge: Reforge? = null
        var usedStone = false
        if (menu.getCaptiveItems(player).size == 2) {
            val stone = getReforgeStone(menu.getCaptiveItems(player)[1])
            if (stone != null) {
                if (stone.targets.any { it.matches(toReforge) }) {
                    reforge = stone
                    usedStone = true
                }
            }
        }

        if (reforge == null) {
            val existing: MutableList<Reforge> = ArrayList()
            if (existingReforge != null) {
                existing.add(existingReforge)
            }
            reforge = getRandomReforge(target, existing)
        }

        if (reforge == null) {
            return
        }

        var cost = 0.0

        if (EconomyManager.hasRegistrations()) {
            cost = plugin.configYml.getDouble("reforge.cost")
            val reforges = getReforges(toReforge)
            cost *= plugin.configYml.getDouble("reforge.cost-exponent").pow(reforges.toDouble())
            if (reforge.requiresStone && reforge.stonePrice != -1) {
                cost = reforge.stonePrice.toDouble()
            }
            cost *= getForPlayer(player).multiplier
            if (!EconomyManager.hasAmount(player, cost)) {
                player.sendMessage(plugin.langYml.getMessage("insufficient-money"))
                if (plugin.configYml.getBool("gui.insufficient-money-sound.enabled")) {
                    player.playSound(
                        player.location,
                        Sound.valueOf(
                            plugin.configYml.getString("gui.insufficient-money-sound.id").uppercase(Locale.getDefault())
                        ),
                        1f, plugin.configYml.getDouble("gui.insufficient-money-sound.pitch").toFloat()
                    )
                }
                return
            }
        }
        var xpCost = plugin.configYml.getInt("reforge.xp-cost")
        val reforges = getReforges(toReforge)
        xpCost *= plugin.configYml.getDouble("reforge.cost-exponent").pow(reforges.toDouble()).toInt()
        xpCost *= getForPlayer(player).multiplier.toInt()
        if (player.level < xpCost) {
            player.sendMessage(plugin.langYml.getMessage("insufficient-xp"))
            if (plugin.configYml.getBool("gui.insufficient-money-sound.enabled")) {
                player.playSound(
                    player.location,
                    Sound.valueOf(
                        plugin.configYml.getString("gui.insufficient-money-sound.id").uppercase(Locale.getDefault())
                    ),
                    1f, plugin.configYml.getDouble("gui.insufficient-money-sound.pitch").toFloat()
                )
            }
            return
        }
        if (EconomyManager.hasRegistrations()) {
            EconomyManager.removeMoney(player, cost)
        }
        player.level = player.level - xpCost
        player.sendMessage(plugin.langYml.getMessage("applied-reforge").replace("%reforge%", reforge.name))
        incrementReforges(toReforge)
        setReforge(toReforge, reforge)
        if (usedStone) {
            val stone = menu.getCaptiveItems(player)[1]
            stone.itemMeta = null
            stone.amount = 0
            if (plugin.configYml.getBool("gui.stone-sound.enabled")) {
                player.playSound(
                    player.location,
                    Sound.valueOf(plugin.configYml.getString("gui.stone-sound.id").uppercase(Locale.getDefault())),
                    1f, plugin.configYml.getDouble("gui.stone-sound.pitch").toFloat()
                )
            }
        }
        if (plugin.configYml.getBool("gui.sound.enabled")) {
            player.playSound(
                player.location,
                Sound.valueOf(plugin.configYml.getString("gui.sound.id").uppercase(Locale.getDefault())),
                1f, plugin.configYml.getDouble("gui.sound.pitch").toFloat()
            )
        }
    }
}
